File: 백준/Gold/2110. 공유기 설치/공유기 설치.py
import sys
input = sys.stdin.readline

n, c = map(int, input().split())
house = sorted([int(input()) for _ in range(n)])

l, r = 1, house[-1] - house[0]
ans = 0

while l <= r:
    mid = (l + r) // 2
    
    now = house[0]
    cnt = 1
    for i in range(1, n):
        if house[i] >= now + mid:
            cnt += 1
            now = house[i]
    
    if cnt >= c:
        l = mid + 1
        ans = mid
    else:
        r = mid - 1

print(ans)

File: 백준/Gold/2110. 공유기 설치/공유기 설치.java
import java.util.*;
import java.io.*;

public class Main {

    static int n, c;
    static int[] X;

    static int check(long target) {

        int cnt = 1;
        int now = X[0];
        for (int i = 1; i < n; i++) {
            long diff = Math.abs(now - X[i]);
            if (target <= diff) {
                cnt++;
                now = X[i];
            }

        }

        return cnt;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        c = Integer.parseInt(st.nextToken());
        X = new int[n];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            X[i] = Integer.parseInt(st.nextToken());
        }

        Arrays.sort(X); // O(N log N)

        long left = 0;
        long right = 1_000_000_001;
        // O(N log N)
        while (left < right) {
            long mid = (left + right) / 2;

            int count = check(mid);

            if (count >= c) {
                left = mid+1;
            } else {
                right = mid;
            }
        }
        bw.write(String.valueOf(left-1));
        bw.flush();
        bw.close();
        br.close();
    }
}